orkamv_visibility_toggle_controller.py

```diff
@@ -0,0 +1,209 @@
+    dlg: Optional[OrkamvVisibilityTogglePluginDialog] = None
+
+    selected_layer_group: str = None
+    selected_groups_hash: Dict[str, List[str]] = {}
+
+    def __init__(self, iface: QgisInterface):
+        self.iface = iface
+        # initialize plugin directory
+        self.plugin_dir = os.path.dirname(__file__)
+        # initialize locale
+        locale = QSettings().value('locale/userLocale')[0:2]
+        locale_path = os.path.join(
+            self.plugin_dir,
+            'i18n',
+            'OrkamvDataApiPlugin_{}.qm'.format(locale))
+
+        if os.path.exists(locale_path):
+            self.translator = QTranslator()
+            self.translator.load(locale_path)
+            QCoreApplication.installTranslator(self.translator)
+
+    # noinspection PyMethodMayBeStatic
+    def tr(self, message):
+        """Get the translation for a string using Qt translation API.
+
+        We implement this ourselves since we do not inherit QObject.
+
+        :param message: String for translation.
+        :type message: str, QString
+
+        :returns: Translated version of message.
+        :rtype: QString
+        """
+        # noinspection PyTypeChecker,PyArgumentList,PyCallByClass
+        return QCoreApplication.translate('OrkamvDataApiPlugin', message)
+
+    def run(self):
+        self.dlg = OrkamvVisibilityTogglePluginDialog()
+
+        self.refresh_groups_hash()
+        self.populate_layer_combo()
+        self.init_group_selection()
+        self.connect_events()
+
+        # show the dialog
+        self.dlg.show()
+
+        # Run the dialog event loop
+        result = self.dlg.exec_()
+
+        # See if OK was pressed
+        if result:
+            # Do something useful here - delete the line containing pass and
+            # substitute with your code.
+            pass
+
+    def connect_events(self):
+        QgsProject.instance().layersAdded.connect(self.reset)
+        QgsProject.instance().layersRemoved.connect(self.reset)
+        self.dlg.layer_select_combo.activated.connect(self.on_combo_change)
+
+    def reset(self):
+        self.refresh_groups_hash()
+        self.populate_layer_combo()
+        config = self.get_groups_config()
+        self.populate_group_selection(config)
+
+    def refresh_groups_hash(self):
+        layer_groups = self.get_layer_groups()
+        layer_group_names = [layer_group.name() for layer_group in layer_groups]
+        for group in layer_groups:
+            if group.name() not in self.selected_groups_hash:
+                group_config = group.customProperty(LAYER_GROUPS_PROP_NAME)
+                self.selected_groups_hash[group.name()] = [g['title'] for g in group_config]
+        to_delete = []
+        for key in self.selected_groups_hash:
+            if key not in layer_group_names:
+                to_delete.append(key)
+        for k in to_delete:
+            if k in self.selected_groups_hash:
+                del self.selected_groups_hash[k]
```

Comment:
```suggestion
        layer_groups = self.get_layer_groups()
        self.selected_groups_hash = {
          group.name(): [
            g['title'] for g in group.customProperty(LAYER_GROUPS_PROP_NAME)
          ] for group in layer_groups
        }
```
Is something like this possible? recreating the hash should not be more much more expensive than adding and deleting, maybe even cheaper. This would be easier to maintain.

For me the only reason to keep this would be that the reference to the hash has to be kept alive because of external dependencies.

Reply:
Good point. Currently, I would keep it as is, since this state was successfully tested. But it might be subject of refactoring the next time I'll be working on the project.

Reply:
maybe open a PR for this?

Reply:
done https://github.com/terrestris/orkamv-data-api-plugin/pull/14